core/common/src/main/java/alluxio/conf/DeprecatedKeyChecker.java

```diff
@@ -0,0 +1,131 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.conf;
+
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+
+import javax.annotation.Nullable;
+
+/**
+ * This annotation checker should be used to determine whether a {@link PropertyKey} or
+ * {@link alluxio.conf.PropertyKey.Template} has a given annotation.
+ *
+ * The class is mainly useful to check for {@link Deprecated} annotations on property keys and
+ * return the associated message, if any.
+ */
+public class DeprecatedKeyChecker {
+  private static final Class<PropertyKey.Template> TEMPLATE_CLASS =
+      PropertyKey.Template.class;
+  private static final Class<PropertyKey> KEY_CLASS = PropertyKey.class;
+
+  private Map<PropertyKey, Deprecated> mAnnotatedKeys;
+  private Map<PropertyKey.Template, Deprecated> mAnnotatedTemplates;
+  private AtomicBoolean mInitialized = new AtomicBoolean(false);
+
+  /**
+   * Create a new instance of {@link DeprecatedKeyChecker}.
+   */
+  public DeprecatedKeyChecker() { }
+
+  /**
+   * Given a class to search, a field type, and an annotation type will return a map of all
+   * fields which are marked with the given annotation to the instance of the annotation.
+   *
+   * @param searchType the class to search through of the given type
+   * @param <T> The type of the field to retrieve
+   * @return a map of all fields within {@code searchType} class of the type T that are annotated
+   *         with the {@link Deprecated} annotation
+   */
+  private static <T> Map<T, Deprecated> populateAnnotatedKeyMap(Class<T> searchType) {
+    Map<T, Deprecated> annotations = new HashMap<>();
+    // Iterate over all fields in the class
+    for (Field field : searchType.getDeclaredFields()) {
+      // If the field isn't equal to the class type, skip it
+      if (!field.getType().equals(searchType)) {
+        continue;
+      }
+
+      Deprecated keyAnnotation = field.getAnnotation(Deprecated.class);
+
+      try {
+        // Field#get parameter can be null if retrieving a static field (all PKs are static)
+        // See https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Field.html
+        // This also works with Template enums
+        T key = searchType.cast(field.get(null));
+        if (keyAnnotation != null) {
+          annotations.put(key, keyAnnotation);
+        }
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    return annotations;
+  }
+
+  private Deprecated getKeyAnnotation(PropertyKey key) {
+    if (!mInitialized.get()) {
```

Comment:
this looks like a lazy initialization. Any particular reason to use lazy initialization rather than static init? Are we concerned about the class init order?

Reply:
Yes, it is lazy initialization. I am concerned about order because we can't retrieve the keys on the fields in the class haven't been instantiated yet. So we initialize lazily to ensure that the map can properly be populated with all keys.